integration/networking/port_mapping_linux_test.go

```diff
@@ -0,0 +1,313 @@
+
+	inspect := container.Inspect(ctx, t, c, serverId)
+	hostPort := inspect.NetworkSettings.Ports["80/tcp"][0].HostPort
+
+	resp, err := http.Get("http://[::1]:" + hostPort)
+	assert.NilError(t, err)
+	assert.Check(t, is.Equal(resp.StatusCode, 404))
+}
+
+// TestAccessPublishedPortFromLocalHost checks whether published ports are
+// accessible from the host netns.
+func TestAccessPublishedPortFromLocalHost(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows")
+	skip.If(t, testEnv.IsRootless())
+
+	ctx := setupTest(t)
+
+	assert.NilError(t, exec.Command("sysctl", "net.ipv6.conf.all.disable_ipv6=0").Run())
+	out, err := exec.Command("ip", "addr", "add", "fd85:1234:4321:cafe::1/64", "dev", "eth0").CombinedOutput()
+	if err != nil && !strings.Contains(string(out), "address already assigned") {
+		assert.NilError(t, err, string(out))
+	}
+
+	testcases := []struct {
+		name        string
+		daemonFlag  string
+		clientIface string
+		ipv6        bool
+	}{
+		{
+			name:        "userland-proxy=true/IPv4/loopback address",
+			daemonFlag:  "--userland-proxy=true",
+			clientIface: "lo",
+		},
+		{
+			name:        "userland-proxy=true/IPv4/local IP address",
+			daemonFlag:  "--userland-proxy=true",
+			clientIface: "eth0",
+		},
+		{
+			name:        "userland-proxy=false/IPv4/loopback address",
+			daemonFlag:  "--userland-proxy=false",
+			clientIface: "lo",
+		},
+		{
+			name:        "userland-proxy=false/IPv4/local IP address",
+			daemonFlag:  "--userland-proxy=false",
+			clientIface: "eth0",
+		},
+		{
+			name:        "userland-proxy=true/IPv6/local IP address",
+			daemonFlag:  "--userland-proxy=true",
+			clientIface: "eth0",
+			ipv6:        true,
+		},
+		{
+			name:        "userland-proxy=false/IPv6/local IP address",
+			daemonFlag:  "--userland-proxy=false",
+			clientIface: "eth0",
+			ipv6:        true,
+		},
+	}
+
+	for tcID, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := testutil.StartSpan(ctx, t)
+
+			clientAddr, err := getIfaceAddress(t, tc.clientIface, tc.ipv6)
+			assert.NilError(t, err)
+
+			d := daemon.New(t)
+			d.StartWithBusybox(ctx, t, tc.daemonFlag)
+			defer d.Stop(t)
+
+			c := d.NewClientT(t)
+			defer c.Close()
+
+			bridgeName := fmt.Sprintf("nat-lo-%d", tcID)
+			bridgeOpts := []func(options *networktypes.CreateOptions){
+				network.WithDriver("bridge"),
+				network.WithOption("com.docker.network.bridge.name", bridgeName),
+			}
+			if tc.ipv6 {
+				bridgeOpts = append(bridgeOpts,
+					network.WithIPv6(),
+					network.WithIPAM("fdf1:a844:380c:b247::/64", "fdf1:a844:380c:b247::1"))
+			}
+
+			network.CreateNoError(ctx, t, c, bridgeName, bridgeOpts...)
+			defer network.RemoveNoError(ctx, t, c, bridgeName)
+
+			serverPort := strconv.Itoa(1234 + tcID)
+			serverID := container.Run(ctx, t, c,
+				container.WithName(sanitizeCtrName(t.Name()+"-server")),
+				container.WithExposedPorts("80/tcp"),
+				container.WithPortMap(nat.PortMap{"80/tcp": {{HostPort: serverPort}}}),
```

Comment:
As the mapping doesn't include a `HostIP`, the container's port 80 should be accessible via any host address (apart from `--userland-proxy=false --ipv6=false`, which will be IPv4 only).

If that's the intention, perhaps it'd be worth checking access via each of the host addresses in each test... dropping the last component of the test-config so that, for each combo of userland-proxy and IPv6-enabled, there's an http request to each of eth0 and lo's v4 and v6 addresses.

But it might also be worth testing that, with a host address in the port mapping, it's not possible to access the http server via other host addresses?